refactor: TypeScript ~4.8.4 and ESLint deps upgrades and initial type fixes

leotm opened the pull request:
**Description**

This is the initial effort of: https://github.com/MetaMask/metamask-mobile/issues/5993

_Refer to our pinned Google Sheet in Mobile Comms Practise for remaining initial type fixes done_

TODO
- [x] Upgrade to TypeScript <s>5</s> ~4.8.4
- [x] Fix current ESLint warning i.e. upgrade @typescript-eslint/parser to support TypeScript version
- [x] Upgrade `@types` devDeps (and add missing i18n-js)
- [x] Upgrade various ESLint devDeps keeping linting errors 0
- [x] Fix lint warning - remove dupe test
- [x] Doc current @react-native-community lint config
- [x] Convert prev files in https://github.com/MetaMask/metamask-mobile/pull/5975 from JS to TS
- [x] Fix initial set of type errors (ref: our internal Google Sheet pinned in Slack channel)

**Screenshots/Recordings**

_If applicable, add screenshots and/or recordings to visualize the before and after of your change_

**Issue**

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

leotm committed:
TypeScript 5 ESLint upgrade and initial type fixes

leotm commented on package.json:
```diff
@@ -354,7 +355,6 @@
     "appium": "^1.22.3",
     "assert": "1.4.1",
     "babel-core": "7.0.0-bridge.0",
-    "babel-eslint": "10.1.0",
```
untracked here since this is already included and kept in-sync in `@react-native-community/eslint-config`

leotm replied:
was also included in `eslint-config-react-native`
now removed in https://github.com/MetaMask/metamask-mobile/pull/6080#discussion_r1153521045
since unused and no longer maintained

leotm commented on app/__mocks__/react-native-i18n.ts:
```diff
@@ -1,5 +1,3 @@
-// @flow
```
Flow officially deprecated in favour or TS

leotm replied:
reference: `tsconfig.json` introduced since https://github.com/facebook/react-native/tree/0.71-stable/template
(replacing now deprecated https://github.com/react-native-community/react-native-template-typescript)

and spotted few mo ago in https://github.com/leotm/react-native-template-new-architecture/commit/ce344615cc2bbb54b2c6e6ae70253e302d63a1d6

leotm commented on app/components/UI/Identicon/index.tsx:
```diff
@@ -50,35 +63,7 @@ const Identicon = React.memo((props) => {
-};
-
-const mapStateToProps = (state) => ({
+const mapStateToProps = (state: any) => ({
```
our 87 untyped state `any`s can be properly typed as follow-up
after reaching 0 type errors (then likely also time to upg to RTK)

leotm commented on app/components/UI/Identicon/index.tsx:
```diff
@@ -1,21 +1,34 @@
 import { connect } from 'react-redux';
 import { useTheme } from '../../../util/theme';
 
+interface IdenticonProps {
+  diameter: number;
+  address: string;
+  customStyle?: ImageStyle;
+  noFadeIn?: boolean;
+  useBlockieIcon: boolean;
+}
+
```
these can be extracted to a .types.ts and exported

leotm committed:
Add back test and rename both

leotm committed:
Update ESLint config comment

Gudahtt commented on package.json:
```diff
@@ -395,7 +394,7 @@
     "rn-nodeify": "10.0.1",
     "stack-beautifier": "1.0.2",
-    "typescript": "^4.4.2",
+    "typescript": "^5.0.2",
```
The MetaMask shared ESLint config isn't compatible with v5 yet, it's still using `~4.8.4`. Could this update be delayed? Otherwise this will be introducing peer dependency warnings and potentially other incompatibilities.

leotm replied:
i [see](https://github.com/search?q=org%3AMetaMask%20~4.8.4&type=code) what you mean 🙈 sure ^ done in https://github.com/MetaMask/metamask-mobile/pull/6080/commits/daad80ea23cfa3e5f0625324d977f035e7682b67

also brings us from _Found 571 errors in 91 files_ to _Found 569 errors in 90 files as a bonus_

Gudahtt replied:
Nice, thanks!

leotm committed:
Downgrade TypeScript from ^5.0.2 to ~4.8.4

As @Gudahtt pointed out in:
https://github.com/MetaMask/metamask-mobile/pull/6080#discussion_r1155902095

NicolasMassart reviewed (dismissed):
Looks good to me. Thanks @leotm for the explanation of the changes.

NicolasMassart commented:
@leotm I will approve again as soon as we test all this with new Node version. Otherwise looks good to me.

leotm commented on app/util/number/index.test.ts:
```diff
@@ -784,9 +784,6 @@ describe('Number utils :: isZeroValue', () => {
   });
-  it('returns true for 0x0', () => {
-    expect(isZeroValue(0x0)).toBe(true);
-  });
```
updated in https://github.com/MetaMask/metamask-mobile/pull/6080/commits/4219cc7fc3dc0d4a28ba43c2b7d2fc4241fc096f after earlier discussion with mr @NicolasMassart

leotm committed:
Merge remote-tracking branch 'upstream/main' into typescript-5-eslint-upgrade-initial-fixes

leotm committed:
Update Yarn.lock after conflict resolution

leotm commented:
> @leotm I will approve again as soon as we test all this with new Node version. Otherwise looks good to me.

updated in https://github.com/MetaMask/metamask-mobile/commit/833dea8b1ab5bb3094f4fb67b400e693e063429e#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L493-R493

leotm committed:
Bump CocoaPods from 1.11.3 to 1.12.0

Unsure why Git autoresolved to 1.11.3 nor flagged as conflict

leotm committed:
Revert "Bump CocoaPods from 1.11.3 to 1.12.0"

Nevermind, this is/should only be done in RN upg branch 0.71.6+

This reverts commit 4354a9eeff0f41a87e60b62da197a686006c1565.

leotm commented on app/component-library/components-temp/CustomSpendCap/CustomInput/CustomInput.styles.ts:
```diff
@@ -31,7 +31,7 @@ const styleSheet = (params: { theme: Theme }) => {
       marginRight: 16,
       color: colors.text.default,
-      ...typography.BodyMD,
+      ...typography.sBodyMD,
```
checking further in Storybook:

before (TypeError, TypeStyle not being applied)

<img width="1728" alt="Screenshot 2023-04-06 at 15 04 13" src="https://user-images.githubusercontent.com/1881059/230401910-e02b7574-1cdc-450f-a257-593e4dc3763c.png">

after (TypeError fixed, TypeStyle now being applied, confirmed in snapshot diff)

<img width="1728" alt="Screenshot 2023-04-06 at 15 04 23" src="https://user-images.githubusercontent.com/1881059/230402292-d14de63f-b3fd-46fd-82c4-35c6ba6556c6.png">

presuming this slight change is what we want @brianacnguyen?

if we're to use this component in future

brianacnguyen replied:
hey @leotm that is incorrect. the tokens are used for both mobile and extension, where extension does have a larger screen use case (hence the s/m prefix). for mobile, we're only using smaller screen tokens so all prefixes are removed. BodyMd is basically just sBodyMd on mobile

leotm replied:
thanks again for our call today 👊

brianacnguyen commented on app/component-library/components/Badges/Badge/variants/BadgeNetwork/BadgeNetwork.types.ts:
```diff
@@ -23,8 +23,9 @@ export interface BadgeNetworkProps extends Omit<BadgeBaseProps, 'children'> {
+   * Resolved to "?" if undefined or null.
    */
-  name: string;
+  name?: string;
```
why did this get updated to optional?

brianacnguyen commented on app/component-library/components/Buttons/Button/variants/ButtonLink/ButtonLink.types.ts:
```diff
@@ -19,4 +20,8 @@ export interface ButtonLinkProps extends Omit<ButtonBaseProps, 'labelColor'> {
+  /**
+   * Text component style.
+   */
+  style?: TextStyle;
```
ButtonBase should already have the style prop. Why is this needed?

brianacnguyen reviewed (commented):
Please address comments

leotm committed:
Revert ButtonLink.types.ts TextStyle addition

After call with @brianacnguyen

leotm committed:
Remove Button style from ContractBoxBase

After call with @brianacnguyen

leotm committed:
Merge remote-tracking branch 'upstream/main' into typescript-5-eslint-upgrade-initial-fixes

- Resolve package.json
- Yarn.lock
- BadgeNetwork.types.ts
- app/util/number/index.test.ts

Otherwise regen'ing Yarn and Pod lockfiles causing big diff changes

leotm committed:
Update Yarn lockfile (from prev conflict res)

leotm commented:
@SocketSecurity ignore @typescript-eslint/type-utils@5.57.0
@SocketSecurity ignore @typescript-eslint/utils@5.57.0
@SocketSecurity ignore comment-parser@1.3.1
@SocketSecurity ignore natural-compare-lite@1.4.0

https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=13183551218
- _Add a author field or contributors array to package.json._
  - @typescript-eslint/type-utils@5.57.0 pr to pr @typescript-eslint/eslint-plugin
  - @typescript-eslint/utils@5.57.0 pr to pr @typescript-eslint/eslint-plugin
- @typescript-eslint/eslint-plugin depends on unmaintained natural-compare-lite
  -  likely follow-up w bump to 5.59.2 which may pos resolve (along w @typescript-eslint/parser)
- unmaintained comment-parser unused nor in yarn.lock :suspect: shouldn't be flagged
  - apaz sourced from @metamask/eslint-config@11.1.0 which depends on @metamask/eslint-config-typescript so can be untracked in package.json

sethkfman committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

leotm committed:
Merge remote-tracking branch 'upstream/main' into typescript-5-eslint-upgrade-initial-fixes

leotm committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

leotm committed:
Merge remote-tracking branch 'upstream/main' into typescript-5-eslint-upgrade-initial-fixes

leotm committed:
Fix lint error @typescript-eslint/no-loss-of-precision

Keep mock account balances consistent as hex strings

leotm commented:
`fs` access for static analysis
@SocketSecurity ignore @eslint/eslintrc@2.0.3
@SocketSecurity ignore eslint-plugin-jsdoc@39.9.1
@SocketSecurity ignore get-tsconfig@4.5.0
_via @metamask/eslint-config@11.1.0_
_via eslint-import-resolver-typescript@3.5.4_

No contributors or author data
@SocketSecurity ignore @eslint/js@8.42.0
_via @metamask/eslint-config@11.1.0_

No ReadMe
@SocketSecurity ignore @pkgr/utils@2.3.1
_via eslint-import-resolver-typescript@3.5.4_

https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=14013188899

NicolasMassart reviewed (dismissed):
Looks good to me.

leotm commented:
@SocketSecurity ignore @eslint/eslintrc@2.0.3
@SocketSecurity ignore eslint-plugin-jsdoc@39.9.1
@SocketSecurity ignore get-tsconfig@4.5.0
@SocketSecurity ignore @eslint/js@8.42.0
@SocketSecurity ignore @pkgr/utils@2.3.1

leotm commented:
waiting on RN upg PR below before merging
- https://github.com/MetaMask/metamask-mobile/pull/6220

leotm committed:
Merge remote-tracking branch 'upstream/main' into typescript-5-eslint-upgrade-initial-fixes

leotm commented:
@SocketSecurity ignore semver@4.3.6
@SocketSecurity ignore semver@5.3.0
@SocketSecurity ignore semver@5.4.1
@SocketSecurity ignore semver@5.7.1
@SocketSecurity ignore semver@6.3.0
@SocketSecurity ignore semver@7.0.0
@SocketSecurity ignore semver@7.3.5
@SocketSecurity ignore semver@7.3.7
@SocketSecurity ignore semver@7.3.8
@SocketSecurity ignore semver@7.5.0
@SocketSecurity ignore semver@7.5.1
mildCVE via too many sources from package.json to name

@SocketSecurity ignore tapable@2.2.1
unmaintained, via eslint-import-resolver-typescript@3.5.4

leotm committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

leotm commented:
prev `semver` pkg issues resolved by updating branch with `main`

https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=14556428146:

filesystem access expected for static analysis (eslint, ts)
@SocketSecurity ignore eslint-plugin-jsdoc
@SocketSecurity ignore get-tsconfig

unmaintained
@SocketSecurity ignore react-native-vector-icons
@SocketSecurity ignore tapable
ideally migrate to maintained pkg

No README
@SocketSecurity ignore @pkgr/utils

leotm commented:
prev `semver` pkg issues resolved by updating branch with `main`

https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=14556428146:

filesystem access expected for static analysis (eslint, ts)
@SocketSecurity ignore eslint-plugin-jsdoc
@SocketSecurity ignore get-tsconfig

unmaintained (ideally migrate to maintained pkg)
@SocketSecurity ignore react-native-vector-icons
@SocketSecurity ignore tapable

No README
@SocketSecurity ignore @pkgr/utils

leotm commented:
_prev `semver` pkg issues resolved by updating branch with `main`_

---

@SocketSecurity ignore eslint-plugin-jsdoc
@SocketSecurity ignore get-tsconfig

@SocketSecurity ignore react-native-vector-icons
@SocketSecurity ignore tapable

@SocketSecurity ignore @pkgr/utils

---

https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=14556428146

filesystem access expected for static analysis (eslint, ts)
- eslint-plugin-jsdoc
- get-tsconfig

unmaintained (ideally migrate to maintained pkg)
- react-native-vector-icons
- tapable

No README
- @pkgr/utils

leotm commented:
@SocketSecurity ignore eslint-plugin-jsdoc
@SocketSecurity ignore get-tsconfig
@SocketSecurity ignore react-native-vector-icons
@SocketSecurity ignore tapable
@SocketSecurity ignore @pkgr/utils

_prev `semver` pkg issues resolved by updating branch with `main`_

https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=14556428146
- filesystem access expected for static analysis (eslint, ts)
  - eslint-plugin-jsdoc
  - get-tsconfig
- unmaintained (ideally migrate to maintained pkg)
  - react-native-vector-icons
  - tapable
- No README
  - @pkgr/utils

leotm commented:
nb
- https://socket.dev/npm/package/@pkgr/utils/overview/2.3.1 Supply Chain Security: 76 (supply chain risk)
  - https://socket.dev/npm/issue/obfuscatedRequire severity: high
- https://socket.dev/npm/package/react-native-vector-icons Supply Chain Security: 42 (critical)
  - https://socket.dev/npm/issue/obfuscatedRequire severity: high
- https://socket.dev/npm/package/tapable Supply Chain Security: 100 ✅
- https://socket.dev/npm/package/eslint-plugin-jsdoc Supply Chain Security: 88 ✅
- https://socket.dev/npm/package/get-tsconfig Supply Chain Security: 98 ✅

leotm committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

tommasini reviewed (dismissed):
LGTM!

leotm committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

Conflicts resolved
- app/components/UI/Identicon/index.tsx (pair reviewed w @tommasini)
- package.json
    @typescript-eslint/eslint-plugin: ^5.57.0
    @typescript-eslint/parser: ^5.57.0
    @walletconnect/types: ^2.9.0
    typescript: ~4.8.4

leotm committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

leotm commented:
https://github.com/MetaMask/metamask-mobile/pull/6080/checks?check_run_id=15201697183

@SocketSecurity ignore @eslint/js@8.44.0
@SocketSecurity ignore @eslint/eslintrc@2.1.0

filesystem access expected for static analysis
- @eslint/js@8.44.0

No contributors or author data (package.json)
- @eslint/eslintrc@2.1.0

leotm committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes

Conflict resolved
- yarn.lock (via running yarn)

Gudahtt commented on package.json:
```diff
@@ -343,7 +343,7 @@
     "@ethersproject/providers": "^5.7.2",
     "@lavamoat/allow-scripts": "^1.0.6",
-    "@metamask/eslint-config": "^7.0.0",
+    "@metamask/eslint-config": "^11.1.0",
```
This is bringing in a few peer dependency warnings:

```
warning " > @metamask/eslint-config@11.1.0" has incorrect peer dependency "eslint@^8.27.0".            
warning " > @metamask/eslint-config@11.1.0" has unmet peer dependency "eslint-config-prettier@^8.5.0".
warning " > @metamask/eslint-config@11.1.0" has unmet peer dependency "eslint-plugin-jsdoc@^39.6.2".    
warning " > @metamask/eslint-config@11.1.0" has incorrect peer dependency "eslint-plugin-prettier@^4.2.1".
warning " > @metamask/eslint-config-typescript@7.0.1" has incorrect peer dependency "@metamask/eslint-config@^7.0.0".
warning " > @metamask/eslint-config-typescript@7.0.1" has incorrect peer dependency "@typescript-eslint/eslint-plugin@^4.20.
0".                                                                                                                         
warning " > @metamask/eslint-config-typescript@7.0.1" has incorrect peer dependency "@typescript-eslint/parser@^4.20.0".
warning " > @metamask/eslint-config-typescript@7.0.1" has incorrect peer dependency "eslint@^7.23.0".
```

Gudahtt replied:
I tried bringing in these updates locally and they caused _a lot_ of lint errors :grimacing: Downgrading to v9 had a lot fewer issues, but there was one remaining peer dependency warning about the TypeScript plugin that has no easy resolution (it'd require downgrading TypeScript, which would be nice to avoid doing)

leotm replied:
thanks again!

Gudahtt commented on package.json:
```diff
@@ -398,12 +399,11 @@
     "eslint-plugin-prettier": "^3.3.1",
     "eslint-plugin-react": "7.16.0",
-    "eslint-plugin-react-native": "3.7.0",
+    "eslint-plugin-react-native": "4.0.0",
```
Could we use a `^` range here instead, or does this need to be pinned to this specific version?

Gudahtt committed:
Downgrade ESLint config to v9 and fix peer dependency warnings

leotm committed:
Bump eslint-plugin-react-native 4.0.0 to ^4.0.0

leotm commented:
@SocketSecurity ignore @metamask/eslint-config@9.0.0
@SocketSecurity ignore comment-parser@1.2.4
@SocketSecurity ignore regextras@0.8.0
@SocketSecurity ignore eslint-plugin-jsdoc@36.1.1

https://github.com/MetaMask/metamask-mobile/runs/15326569677

filesystem access expected for static analysis
- eslint-plugin-jsdoc@36.1.1

new author
- metamask/eslint-config@9.0.0

unmaintained
- comment-parser@1.2.4
- regextras@0.8.0

Gudahtt reviewed (approved):
LGTM!

Cal-L committed:
Merge branch 'main' into typescript-5-eslint-upgrade-initial-fixes
